### PR TITLE
Fix batchnorm problem with sparse matrices when fix_gamma=True

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
As title.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Fix fallback logic for batchnorm operator
- [x] Block usage of fix_gamma=True when input is sparse
- [x] Unit tests

### CODE REVIEW

**Comment by haojin2** [comment]
@marcoabreu Here's the fix.

**Comment by haojin2** [comment]
@eric-haibin-lin @zheng-da

File: src/operator/nn/batch_norm.cc [hunk 1]
```diff
@@ -452,19 +453,56 @@ static inline bool BatchNormStorageType(const nnvm::NodeAttrs &attrs,
                                         std::vector<int> *out_attrs) {
   CHECK_EQ(in_attrs->size(), 5);
   CHECK_EQ(out_attrs->size(), 3);
+  const BatchNormParam &param = nnvm::get<BatchNormParam>(attrs.parsed);
+
+  if ((common::ContainsStorageType(*in_attrs, kRowSparseStorage) ||
+       common::ContainsStorageType(*in_attrs, kCSRStorage)) &&
+      param.fix_gamma) {
+    LOG(FATAL) << "fix_gamma=True is not supported for sparse ndarrays";
```
Comment (marcoabreu) [inline src/operator/nn/batch_norm.cc]: Could you please create a GitHub issue as feature request and link it in this message so users are able to follow up?

We should also add this to the known problems for the next release. @Roshrini do we already have a release manager for 1.3?

File: tests/python/unittest/test_sparse_operator.py [hunk 2]
```diff
@@ -2098,6 +2098,77 @@ def check_scatter_ops(name, shape, lhs_stype, rhs_stype, forward_mxnet_call, for
                           lambda l, r: l + r,
                           rhs_is_scalar=True, verbose=False, density=0.5)
 
+
+@with_seed()
```
Comment (roywei) [inline tests/python/unittest/test_sparse_operator.py]: It will be great to add a reference to test_batchnorm_training in test_operator.py, and note this is to test row_sparse fallback correctness. Just to avoid confusion for other developers who work on batchnorm.

**Comment by haojin2** [comment]
@marcoabreu I've already tagged the related persons, waiting for their reviews.

File: src/operator/nn/batch_norm.cc (hunk 1 above)
Comment (haojin2) [reply src/operator/nn/batch_norm.cc]: Done

File: src/operator/nn/batch_norm.cc [hunk 3]
```diff
@@ -452,19 +453,66 @@ static inline bool BatchNormStorageType(const nnvm::NodeAttrs &attrs,
...
+    LOG(FATAL) << "fix_gamma=True is not supported for sparse ndarrays. Tracked at #11647";
+  }
+  for (int& v : *in_attrs)
+    if (v == - 1) v = kDefaultStorage;
+  bool dispatched = false;
+  if (!dispatched && (common::ContainsStorageType(*in_attrs, kRowSparseStorage) ||
+                      common::ContainsStorageType(*in_attrs, kCSRStorage))) {
+    dispatched = dispatch_fallback(out_attrs, dispatch_mode);
+  }
+#if MXNET_USE_MKLDNN == 1
```
Comment (anirudh2290) [inline src/operator/nn/batch_norm.cc]: do we need this if else here. MKLDNNStorageType is handling both cases.

Comment (haojin2) [reply src/operator/nn/batch_norm.cc]: MKLDNNStorageType is not defined if MXNET_USE_MKLDNN=0

Comment (anirudh2290) [reply src/operator/nn/batch_norm.cc]: okay.

**Comment by haojin2** [comment]
@marcoabreu @anirudh2290 I'll be creating another PR for this with the same changes as there's no way for me to re-trigger the build due to the CI bug.
